storage/scratchpad/src/sparse_merkle/mod.rs

```diff
@@ -509,6 +509,393 @@ where
         ret
     }
 
+    /// The batch version of `serial_update`. The function will return a Sparse Merkle Tree of the final state.
+    /// This method is optmized to avoid repetitive common path updating and parallelize the processing as much
+    /// as possible.
+    pub fn batch_update(
+        &self,
+        updates: Vec<Vec<(HashValue, &V)>>,
+        proof_reader: &impl ProofRead<V>,
+    ) -> Result<Self, UpdateError> {
+        // Flatten, dedup and sort the updates with a btree map since the updates between different
+        // versions may overlap on the same address in which case the latter always overwrites.
+        let kvs = updates
+            .into_iter()
+            .flatten()
+            .collect::<BTreeMap<_, _>>()
+            .into_iter()
+            .collect::<Vec<_>>();
+
+        let current_root = self.root_weak();
+        if kvs.is_empty() {
+            Ok(Self::new_with_base(current_root, self))
+        } else {
+            let root = Self::batch_update_subtree(
+                current_root,
+                &kvs[..],
+                /* depth = */ 0,
+                proof_reader,
+            )?;
+            Ok(Self::new_with_base(root, self))
+        }
+    }
+
+    /// Return the index of the first bit that is 1 at the given depth when kvs is lexicographically sorted.
+    fn binary_search<T>(kvs: &[(HashValue, &T)], depth: usize) -> usize {
```

Comment:
`partition()` or `upper_bound` might be a better name.

Reply:
depth keeps increasing by one, should be able to keep partitioning and not do binary search every time.

Reply:
@gelash Could you elaborate? I didn't quite get what you mean by "keep partitioning" here?

Reply:
I am sorry - I misunderstood the algorithm. I somehow thought that not sorting kvs and then passing down the partition of indices that have 0 (or 1) into recursive calls could be more efficient or concise. But after thinking a bit, I like your way of doing this better.